xmtp/src/client.rs

```diff
@@ -142,19 +147,56 @@ where
             .collect())
     }
 
+    /// Fetch Installations from the Network and create unintialized sessions for newly discovered contacts
+    // TODO: Reduce Visibility
     pub async fn refresh_user_installations(&self, user_address: &str) -> Result<(), ClientError> {
+        let self_install_id = key_fingerprint(&self.account.identity_keys().curve25519);
         let contacts = self.get_contacts(user_address).await?;
 
-        let stored_contacts: Vec<StoredInstallation> =
-            self.store.get_contacts(user_address)?.into();
-        println!("{:?}", contacts);
-        for contact in contacts {
-            println!(" {:?} ", contact)
-        }
+        let installation_map = self
+            .store
+            .get_installations(user_address)?
+            .into_iter()
+            .map(|v| (v.installation_id.clone(), v))
+            .collect::<HashMap<_, _>>();
+
+        let new_installs = contacts
+            .iter()
+            .filter(|contact| self_install_id == contact.installation_id())
+            .filter(|contact| !installation_map.contains_key(&contact.installation_id()))
+            .filter_map(|contact| StoredInstallation::new(contact).ok());
+
+        self.store.conn().unwrap().transaction(
+            |transaction_manager| -> Result<(), ClientError> {
+                for install in new_installs {
+                    info!("Saving Install {}", install.installation_id);
+                    let session = self.create_uninitialized_session(&install.get_contact()?)?;
+                    self.store
+                        .insert_or_ignore_install(install, transaction_manager)?;
+                    self.store.insert_or_ignore_session(
+                        StoredSession::try_from(&session)?,
+                        transaction_manager,
+                    )?;
+                }
+
+                Ok(())
+            },
```

Comment:
This is the core of the refresh. The network calls are all done in advance to minimize work being done in the transaction lambda.